### PR TITLE
Add spider for Campos dos Goytacazes/RJ

### PR DESCRIPTION
**Pull request by dannnylo**
In this PR we add the spider to collect data from [Campos dos Goytacazes/RJ gazette](https://www.campos.rj.gov.br/diario-oficial.php)

### CODE REVIEW

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                  file_urls=[url],
+                  is_extra_edition=extra_edition,
+                  municipality_id=self.MUNICIPALITY_ID,
+                  power='executive',
+                  scraped_at=datetime.utcnow(),
+                )
+
+        next_url = response.css('.pagination').xpath("//a[contains(text(), 'Proxima')]/@href").extract_first()
+        if next_url:
+            print(next_url)
```
Comment (giovanisleite) [inline processing/data_collection/gazette/spiders/rj_campos_goytacazes.py]: Remove the print

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import dateparser
+
+from datetime import datetime
+from scrapy import Request
+from gazette.items import Gazette
+from gazette.spiders.base import BaseGazetteSpider
+
+class RjCampoGoytacazesSpider(BaseGazetteSpider):
+    GAZETTE_ELEMENT_CSS = 'ul.ul-licitacoes li'
+    MUNICIPALITY_ID = '3301009'
```
Comment (giovanisleite) [inline processing/data_collection/gazette/spiders/rj_campos_goytacazes.py]: The ```MUNICIPALITY_ID``` property was updated to ```TERRITORY_ID```

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        @scrapes date file_urls is_extra_edition municipality_id power scraped_at
+        """
+
+        for element in response.css(self.GAZETTE_ELEMENT_CSS):
+            url = element.css('a::attr(href)').extract_first()
+            gazette_text = element.css("h4::text").extract_first()
+            extra_edition = gazette_text.startswith('Suplemento')
+
+            date_re = re.search(r'(\d{2} de (.*) de \d{4})', gazette_text)
+            if date_re:
```
Comment (giovanisleite) [inline processing/data_collection/gazette/spiders/rj_campos_goytacazes.py]: Did you see some gazette without date or in a different format?

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        """
+        @url https://www.campos.rj.gov.br/diario-oficial.php?PGpagina=1&PGporPagina=15
+        @returns requests 1
+        @scrapes date file_urls is_extra_edition municipality_id power scraped_at
+        """
+
+        for element in response.css(self.GAZETTE_ELEMENT_CSS):
+            url = element.css('a::attr(href)').extract_first()
+            gazette_text = element.css("h4::text").extract_first()
+            extra_edition = gazette_text.startswith('Suplemento')
```
Comment (giovanisleite) [inline processing/data_collection/gazette/spiders/rj_campos_goytacazes.py]: Nice :fireworks: !

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                  is_extra_edition=extra_edition,
+                  municipality_id=self.MUNICIPALITY_ID,
+                  power='executive',
+                  scraped_at=datetime.utcnow(),
+                )
+
+        next_url = response.css('.pagination').xpath("//a[contains(text(), 'Proxima')]/@href").extract_first()
+        if next_url:
+            print(next_url)
+            yield  Request(response.urljoin(next_url))
```
Comment (giovanisleite) [inline processing/data_collection/gazette/spiders/rj_campos_goytacazes.py]: Mandatory: If I'm not wrong, the default for the callback is None, so you have to declare it ```callback=self.parse```. 

Good to know: A shortcut to create a request without import it that I learned from rennerocha it is ```response.follow```, to this case: ```yield response.follow(response.urljoin(next_url), callback=self.parse)```

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    MUNICIPALITY_ID = '3301009'
+
+    allowed_domains = ['www.campos.rj.gov.br']
+    name = 'rj_campos_goytacazes'
+    start_urls = ['https://www.campos.rj.gov.br/diario-oficial.php?PGpagina=1&PGporPagina=15']
+
+    def parse(self, response):
+        """
+        @url https://www.campos.rj.gov.br/diario-oficial.php?PGpagina=1&PGporPagina=15
+        @returns requests 1
```
Comment (giovanisleite) [inline processing/data_collection/gazette/spiders/rj_campos_goytacazes.py]: ```@returns items 15 15```

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py (hunk 1 above)
Comment (chamun) [reply processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:44]: Addressed at https://github.com/okfn-brasil/querido-diario/pull/235/commits/24283368d8d6104cf3d2c1f041a187c92fce61ee.

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py (hunk 2 above)
Comment (chamun) [reply processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:11]: Addressed at https://github.com/okfn-brasil/querido-diario/pull/235/commits/93e8eefa39fc53803450fca9d97144fcc1fa6949.

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py (hunk 3 above)
Comment (chamun) [reply processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:30]: The first list item of the list has no date.

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py (hunk 6 above)
Comment (chamun) [reply processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:20]: Addressed at https://github.com/okfn-brasil/querido-diario/pull/235/commits/7c34cceb0512098b8071618937344b4d14bbfd59.

**Comment by chamun** [comment]
I have addressed the change requests and update the code from this PR at https://github.com/okfn-brasil/querido-diario/pull/235 as part of #221.

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import re
+import dateparser
+
+from datetime import datetime
+from scrapy import Request
+from gazette.items import Gazette
+from gazette.spiders.base import BaseGazetteSpider
```
Comment (jvanz) [inline processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:7]:
```suggestion
from datetime import datetime
import re

import dateparser
from scrapy import Request

from gazette.items import Gazette
from gazette.spiders.base import BaseGazetteSpider
```

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py (hunk 2 above)
Comment (jvanz) [reply processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:11]: @chamun you need to update this branch with the commit.

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py (hunk 6 above)
Comment (jvanz) [reply processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:20]: @chamun you need to update this branch with the commit.

File: processing/data_collection/gazette/spiders/rj_campos_goytacazes.py (hunk 1 above)
Comment (jvanz) [reply processing/data_collection/gazette/spiders/rj_campos_goytacazes.py:44]: @chamun you need to update this branch with the commit.

**Review by jvanz — CHANGES_REQUESTED** [review]

**Comment by chamun** [comment]
@jvanz I don't have access to this branch so I can't add changes here, that's why I opened #235.

I have addressed @giovanisleite's requests there.

I believe you can close this PR in favor of #235. It contains the original commits by @dannnylo, the changes requests from @giovanisleite and some other fixes I found while running the spider.

**Pull request closed without merging**